packages/backend/src/peripherals/database/TxCountRepository.ts

```diff
@@ -0,0 +1,139 @@
+import { Logger } from '@l2beat/common'
+import { ProjectId, UnixTime } from '@l2beat/types'
+import { TxCountRow } from 'knex/types/tables'
+import _ from 'lodash'
+
+import { BaseRepository } from './shared/BaseRepository'
+import { Database } from './shared/Database'
+
+export interface TxCountRecord {
+  timestamp: UnixTime
+  projectId: ProjectId
+  blockNumber: number
+  count: number
+}
+
+export class TxCountRepository extends BaseRepository {
+  constructor(database: Database, logger: Logger) {
+    super(database, logger)
+
+    /* eslint-disable @typescript-eslint/unbound-method */
+
+    this.add = this.wrapAdd(this.add)
+    this.addMany = this.wrapAddMany(this.addMany)
+    this.deleteAll = this.wrapDelete(this.deleteAll)
+
+    /* eslint-enable @typescript-eslint/unbound-method */
+  }
+
+  async add(record: TxCountRecord) {
+    const knex = await this.knex()
+    const row = toRow(record)
+    await knex('tx_count').insert(row)
+    return `${row.project_id}-${row.block_number}`
+  }
+
+  async addMany(records: TxCountRecord[]) {
+    const knex = await this.knex()
+    const rows = records.map(toRow)
+    await knex('tx_count').insert(rows)
+    return rows.length
+  }
+
+  async findLatestByProject(projectId: ProjectId) {
+    const knex = await this.knex()
+    const row = await knex('tx_count')
+      .where('project_id', projectId.toString())
+      .orderBy('block_number', 'desc')
+      .first()
+
+    return row ? toRecord(row) : undefined
+  }
+
+  async getMissingRangeByProject(projectId: ProjectId) {
+    const knex = await this.knex()
+    const maxBlock = await this.findLatestByProject(projectId)
+    if (!maxBlock) {
+      return []
+    }
+
+    const noNextBlockQuery = knex
+      .select('t1.block_number')
```

Comment:
I really dont think we should rely on knex query builder for such complicated queries. Plain SQL would be much more readable here IMO. @sz-piotr thoughts?

Reply:
I agree, especially since we lose type safety anyway